python/ray/tune/integration/mlflow.py

```diff
@@ -0,0 +1,337 @@
+            config={
+                # define search space here
+                "a": tune.choice([1, 2, 3]),
+                "b": tune.choice([4, 5, 6]),
+                # mlflow configuration
+                "mlflow": {
+                    "experiment_name": "my_experiment",
+                    "tracking_uri": mlflow.get_tracking_uri()
+                }
+            })
+    """
+    if mlflow is None:
+        raise RuntimeError("mlflow has not been installed. Please `pip "
+                           "install mlflow` to use the mlflow_mixin.")
+    func.__mixins__ = (MLFlowTrainableMixin, )
+    return func
+
+
+class MLFlowTrainableMixin:
+    _mlflow = mlflow
+
+    def __init__(self, config: Dict, *args, **kwargs):
+        if not isinstance(self, Trainable):
+            raise ValueError(
+                "The `MLFlowTrainableMixin` can only be used as a mixin "
+                "for `tune.Trainable` classes. Please make sure your "
+                "class inherits from both. For example: "
+                "`class YourTrainable(MLFlowTrainableMixin)`.")
+
+        super().__init__(config, *args, **kwargs)
+        _config = config.copy()
+        try:
+            mlflow_config = _config.pop("mlflow").copy()
+        except KeyError as e:
+            raise ValueError(
+                "MLFlow mixin specified but no configuration has been passed. "
+                "Make sure to include a `mlflow` key in your `config` dict "
+                "containing at least a `tracking_uri` and either "
+                "`experiment_name` or `experiment_id` specification.") from e
+
+        tracking_uri = mlflow_config.pop("tracking_uri", None)
+        if tracking_uri is None:
+            raise ValueError("MLFlow mixin specified but no "
+                             "tracking_uri has been "
+                             "passed in. Make sure to include a `mlflow` "
+                             "key in your `config` dict containing at "
+                             "least a `tracking_uri`")
+        self._mlflow.set_tracking_uri(tracking_uri)
+
+        # First see if experiment_id is passed in.
+        experiment_id = mlflow_config.pop("experiment_id", None)
+        if experiment_id is None:
+            # Check for name.
+            experiment_name = mlflow_config.pop("experiment_name", None)
+            if experiment_name is None:
+                raise ValueError(
+                    "MLFlow mixin specified but no "
+                    "experiment_name or experiment_id has been "
+                    "passed in. Make sure to include a `mlflow` "
+                    "key in your `config` dict containing at "
+                    "least a `experiment_name` or `experiment_id` "
+                    "specification.")
+            # Name exists.
+            experiment = self._mlflow.get_experiment_by_name(experiment_name)
+            experiment_id = experiment.experiment_id
+
+        self.experiment_id = experiment_id
```

Comment:
do we have to set `mlflow.set_experiment` somewhere here? I think we set it in some of the driver functions. Or is this subsumed by `start_run`?

Reply:
No we don't have to call set_experiment, we call create_experiment instead if the experiment does not already exist, and then directly pass in the experiment_id when we are creating the runs.